### PR TITLE
Ui collection detail serializers 102

### PR DESCRIPTION
**Pull request by alikins**
Making  return in the format specified at https://github.com/ansible/galaxy-dev/issues/97 and https://github.com/ansible/galaxy-api/pull/75

Fixes: https://github.com/ansible/galaxy-dev/issues/102

### CODE REVIEW

**Comment by alikins** [comment]
(note: I think the openapi spec is out of date with the latest changes at the moment cc87d113188e17c18b8f04375bf0f72904918e41 but will reconcile)

**Comment by alikins** [comment]
@newswangerd 

> duplicate collections are getting returned on the list view now

I think f223573134a1bc430d8ca808cda37a449696084b should fix this

File: galaxy_api/api/ui/serializers/collection.py [hunk 1]
```diff
@@ -50,26 +61,26 @@ class CollectionVersionBaseSerializer(serializers.Serializer):
...
-    created_at = serializers.DateTimeField(source='_created')
 
+    created_at = serializers.DateTimeField(source='_created')
 
 class CollectionLatestVersionSerializer(CollectionVersionBaseSerializer):
     metadata = CollectionMetadataBaseSerializer(source='*')
 
+class CollectionLatestVersionDetailSerializer(CollectionLatestVersionSerializer):
+    docs_blob = serializers.JSONField()
+    contents = serializers.ListField(ContentSerializer())
```
Comment (newswangerd) [inline galaxy_api/api/ui/serializers/collection.py]: I need `contents` to get returned in the list view.

**Comment by alikins** [comment]
@newswangerd I think https://github.com/ansible/galaxy-api/pull/85#issuecomment-530901860 should be fixed with bcfc0d8e034eea7581e0d6b40e8641947bd75d5d

File: galaxy_api/api/ui/serializers/collection.py (hunk 1 above)
Comment (alikins) [reply galaxy_api/api/ui/serializers/collection.py]: And the list view `['latest_version']['contents']` item models are the same as for the collection detail view?  

But the 'contents' in detail view ( ['latest_version']['docs_blob']['contents'] ) is different?

ie, list view `['latest_version']['contents']`:

``` json
            "latest_version": {
                "contents": [
                    {
                        "content_type": "module",
                        "description": null,
                        "name": "qradar_deploy"
                    },
                   " #  ITEMS TRIMMED FOR BREVITY ",
                    {
                        "content_type": "module_utils",
                        "description": null,
                        "name": "qradar"
                    },
                    {
                        "content_type": "httpapi",
                        "description": "HttpApi Plugin for IBM QRadar",
                        "name": "qradar"
                    }
                ],
```

detail view ['latest_content']['contents']:

``` json
    "latest_version": {
        "contents": [
            {
                "content_type": "module",
                "description": null,
                "name": "qradar_deploy"
            },
            " # ITEMS TRIMMED FOR BREVITY",
            {
                "content_type": "module_utils",
                "description": null,
                "name": "qradar"
            },
            {
                "content_type": "httpapi",
                "description": "HttpApi Plugin for IBM QRadar",
                "name": "qradar"
            }
        ],
```

But different from the 'contents' field in detail view ['latest_version']['docs_blob']['contents']

``` json
    "latest_version": {
        "docs_blob": {
            "contents": [
                {
                    "content_name": "qradar_deploy",
                    "content_type": "module",
                    "doc_strings": null,
                    "readme_file": null,
                    "readme_html": null
                },
                " # ITEMS SKIPPED FOR BREVITY",
                {
                    "content_name": "qradar",
                    "content_type": "module_utils",
                    "doc_strings": null,
                    "readme_file": null,
                    "readme_html": null
                },
                {
                    "content_name": "qradar",
                    "content_type": "httpapi",
                    "doc_strings": {
                        "doc": {
                            "author": "Ansible Security Automation Team",
                            "description": [
                                "This HttpApi plugin provides methods to connect to IBM QRadar over a HTTP(S)-based api."
                            ],
                            "filename": "/tmp/tmp5iifq9uu/plugins/httpapi/qradar.py",
                            "httpapi": "qradar",
                            "short_description": "HttpApi Plugin for IBM QRadar",
                            "version_added": "2.9"
                        },
                        "examples": null,
                        "metadata": {
                            "status": [
                                "preview"
                            ],
                            "supported_by": "community"
                        },
                        "return": null
                    },
                    "readme_file": null,
                    "readme_html": null
                }
```

Comment (alikins) [reply galaxy_api/api/ui/serializers/collection.py]: I think e371142d13d8449fd4dae7b90b75855876c6ede1 should add 'contents' to both list and detail views.

File: galaxy_api/api/ui/serializers/collection.py [hunk 2]
```diff
@@ -50,42 +61,46 @@ class CollectionVersionBaseSerializer(serializers.Serializer):
...
 class CollectionListSerializer(_CollectionSerializer):
     def _get_namespace(self, obj):
         name = obj['namespace']
-        return self.context['namespaces'][name]
+        return self.context['namespaces'].get(name, None)
 
 
 class CollectionDetailSerializer(_CollectionSerializer):
+    latest_version = CollectionLatestVersionDetailSerializer(source='*')
+    all_versions = serializers.SerializerMethodField()
```
Comment (newswangerd) [inline galaxy_api/api/ui/serializers/collection.py]: `all_versions` doesn't contain a list of all the versions when the `?version=` param is set.

File: galaxy_api/api/ui/serializers/collection.py [hunk 3]
```diff
@@ -50,42 +61,46 @@ class CollectionVersionBaseSerializer(serializers.Serializer):
...
-    created_at = serializers.DateTimeField(source='_created')
 
+    created_at = serializers.DateTimeField(source='_created')
 
 class CollectionLatestVersionSerializer(CollectionVersionBaseSerializer):
     metadata = CollectionMetadataBaseSerializer(source='*')
 
+class CollectionLatestVersionDetailSerializer(CollectionLatestVersionSerializer):
+    docs_blob = serializers.JSONField()
+    contents = serializers.ListField(ContentSerializer())
```
Comment (newswangerd) [inline galaxy_api/api/ui/serializers/collection.py]: I need `contents` to get returned in the list view as well as the detail view.

Comment (alikins) [reply galaxy_api/api/ui/serializers/collection.py]: Doesn't https://github.com/ansible/galaxy-api/pull/85/commits/e371142d13d8449fd4dae7b90b75855876c6ede1  fix this?

(not sure why, but a comment mentioning that was deleted)

File: galaxy_api/api/ui/serializers/collection.py (hunk 2 above)
Comment (alikins) [reply galaxy_api/api/ui/serializers/collection.py]: Right, `?version` is a param for filtering the list view which doesn't have `all_versions`

File: galaxy_api/api/ui/serializers/collection.py (hunk 3 above)
Comment (alikins) [reply galaxy_api/api/ui/serializers/collection.py]: @newswangerd  https://gist.github.com/alikins/9d224e5748a1d56c92b6eb99616e3a6f should fix this. Example output https://gist.github.com/alikins/9d224e5748a1d56c92b6eb99616e3a6f

**Pull request merged**